site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+an addition.
+
+- Users of the encapsulated protocol would miss out on significant conventional features from their ecosystem if forced to route at a lower layer.
+- Users of the enapsulated protocol would experience a degraded user experience if forced to route at a lower layer.
+- The encapsulated protocol has a significant user base, particularly in the Kubernetes community.
+
+gRPC meets _all_ of these criteria and is therefore, we contend, a strong
+candidate for inclusion in the Gateway APIs.
+
+#### HTTP/2 Cleartext
+
+gRPC allows HTTP/2 cleartext communication (H2C). This is conventionally deployed for
+testing. Many control plane implementations do not support this by default and
+would require special configuration to work properly.
+
+#### Content-Based Routing
+
+While not included in the scope of this initial GEP, a common use case cited for
+routing gRPC is payload-aware routing. That is, routing rules which determine a
+backend based on the contents of the protocol buffer payload.
+
+#### User Experience
+
+The user experience would also degrade significantly if forced to route at the level of HTTP.
+
+- Encoding services and methods as URIs (an implementation detail of gRPC)
+- The [Transfer Encoding header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) for trailers
+- Many features supported by HTTP/2 but not by gRPC, such as
+  - Query parameters
+  - Methods besides `POST`
+  - CORS
+
+
+#### Proxyless Service Mesh
+
+The gRPC library supports proxyless service mesh, a system by which routing
+configuration is received not by an in-line proxy or sidecar proxy but by the client
+itself. Eventually, `GRPCRoute` in the Gateway APIs should support this feature.
+However, to date, there are no HTTP client libraries capable of participating
+in a proxyless service mesh.
+
+--- 
+
+### Cross Serving
+
+A new route type (and especially one encapsulated in an already supported
+protocol) raises questions about precedence of routing rules. Today, there is
+already a complicated algorithm dictating which routing rule to apply to a
+particular HTTP request when two `HTTPRoutes` both apply to the request's
+hostname. What semantics should apply when a `GRPCRoute` and an `HTTPRoute` both apply to
+a request?
+
+We propose that the GRPCRoute's matchers be translated into the equivalent
+HTTPRoute matchers and that the normal HTTPRoute precedence rules apply.
+This translation is mostly straightforward, but one significant deviation between
+GRPCRoute and HTTPRoute poses difficulties: the gRPC method matcher.
+
+gRPC methods are encoded in the URI of an HTTP/2 message in the format
+`/SERVICE/METHOD`. The matcher allows the user to supply these independently, so
+that the user may leave either blank, to be interpreted as an unconditional
+match. Further, this field may either be an exact match or a regex match.
+
+To translate this field to an HTTPRoute path matcher, we propose the following
```

Comment:
Question: Is all of this translation visible to the end-user or is it implementation guidance?
Is that we expect a controller in the cluster that reconciles GRPCRoute into HTTPRoute?

Reply:
This feels like implementation guidance to me - that the GRPCRoute object should be translated into data-path config equivalent to the object. I am a -1 on actually automatically generating an actual HTTPRoute object.

Reply:
+1 on your -1

Reply:
This section has been removed due to a change in merge semantics.